types/wlr_linux_dmabuf_v1.c

```diff
@@ -400,6 +412,257 @@ static void linux_dmabuf_create_params(struct wl_client *client,
 		&buffer_params_impl, params, params_handle_resource_destroy);
 }
 
+static void linux_dmabuf_hints_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zwp_linux_dmabuf_hints_v1_interface
+		linux_dmabuf_hints_impl = {
+	.destroy = linux_dmabuf_hints_destroy,
+};
+
+static bool hints_tranche_init_with_renderer(
+		struct wlr_linux_dmabuf_hints_v1_tranche *tranche,
+		struct wlr_renderer *renderer) {
+	memset(tranche, 0, sizeof(*tranche));
+
+	int drm_fd = wlr_renderer_get_drm_fd(renderer);
+	if (drm_fd < 0) {
+		wlr_log(WLR_ERROR, "Failed to get DRM FD from renderer");
+		return false;
+	}
+
+	struct stat stat;
+	if (fstat(drm_fd, &stat) != 0) {
+		wlr_log_errno(WLR_ERROR, "fstat failed");
+		return false;
+	}
+	tranche->target_device = stat.st_rdev;
+
+	const struct wlr_drm_format_set *formats =
+		wlr_renderer_get_dmabuf_texture_formats(renderer);
+	if (formats == NULL) {
+		wlr_log(WLR_ERROR, "Failed to get renderer DMA-BUF texture formats");
+		return false;
+	}
+
+	if (!wlr_drm_format_set_copy(&tranche->formats, formats)) {
+		wlr_log(WLR_ERROR, "Failed to copy DRM format set");
+		return false;
+	}
+
+	return true;
+}
+
+static bool hints_init_with_renderer(struct wlr_linux_dmabuf_hints_v1 *hints,
+		struct wlr_renderer *renderer) {
+	memset(hints, 0, sizeof(*hints));
+
+	struct wlr_linux_dmabuf_hints_v1_tranche *tranche =
+		calloc(1, sizeof(*tranche));
+	if (tranche == NULL) {
+		return false;
+	}
+
+	if (!hints_tranche_init_with_renderer(tranche, renderer)) {
+		free(tranche);
+		return false;
+	}
+
+	hints->main_device = tranche->target_device;
+	hints->tranches = tranche;
+	hints->tranches_len = 1;
+
+	return true;
+}
+
+static void hints_finish(struct wlr_linux_dmabuf_hints_v1 *hints) {
+	for (size_t i = 0; i < hints->tranches_len; i++) {
+		struct wlr_linux_dmabuf_hints_v1_tranche *tranche = &hints->tranches[i];
+		wlr_drm_format_set_finish(&tranche->formats);
+	}
+	free(hints->tranches);
+}
+
+static bool hints_copy(struct wlr_linux_dmabuf_hints_v1 *dst,
+		const struct wlr_linux_dmabuf_hints_v1 *src) {
+	memset(dst, 0, sizeof(*dst));
+
+	dst->main_device = src->main_device;
+	dst->tranches_len = src->tranches_len;
+
+	dst->tranches = calloc(src->tranches_len,
+		sizeof(struct wlr_linux_dmabuf_hints_v1_tranche));
+	if (dst->tranches == NULL) {
+		return false;
+	}
+
+	for (size_t i = 0; i < src->tranches_len; i++) {
+		dst->tranches[i].target_device = src->tranches[i].target_device;
+		dst->tranches[i].flags = src->tranches[i].flags;
+		if (!wlr_drm_format_set_copy(&dst->tranches[i].formats,
+				&src->tranches[i].formats)) {
+			return false;
```

Comment:
Forgot to free the other `dst->tranches[i].formats` that were allocated and also `dst->tranches`.